protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnConverter.java

```diff
@@ -35,8 +36,16 @@
 
     @VisibleForTesting final ArrayList<MessageUid> uids;
 
-    public UidMsnConverter(Iterator<MessageUid> iterator) {
-        uids = Lists.newArrayList(iterator);
+    public UidMsnConverter() {
+        this.uids = Lists.newArrayList();
+    }
+
+    public synchronized void addAll(List<MessageUid> iterator) {
+        LinkedHashSet<MessageUid> messageUids = new LinkedHashSet<MessageUid>();
```

Comment:
that's a lot of copies : one before calling addAll, one to remove duplicates, and one for the final list.
Why do you use a LinkedHashSet here ?